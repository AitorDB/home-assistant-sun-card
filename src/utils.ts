import { html, TemplateResult } from 'lit-html'
import { Constants } from './constants'
import { TSunCardI18N, TSunCardI18NKeys } from './types'

export class I18N {
  private localization: TSunCardI18NKeys
  private dateFormatter: Intl.DateTimeFormat

  constructor (language: string, use12HourClock: boolean) {
    this.localization = Constants.LOCALIZATION_LANGUAGES[language]

    // https://developer.mozill a.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/DateTimeFormat
    this.dateFormatter = new Intl.DateTimeFormat(language, {
      hour12: use12HourClock,
      timeStyle: 'short'
    })
  }

  public formatDateAsTime (date: Date): string {
    return this.dateFormatter.formatToParts(date).map(({ type, value }) => {
      switch (type) {
        // intentional fallthrough
        case 'hour':
        case 'minute':
        case 'dayPeriod':
        case 'literal':
          return value
        default:
          return ''
      }
    })
    .join('')
  }

  public tr (trKey: string): string {
    return this.getLocalizationElement(this.localization, trKey).toString()
  }

  // Janky recursive logic to handle nested values in i18n json sources
  private getLocalizationElement (localization: TSunCardI18N, trKey: string): string | TSunCardI18N {

    if (trKey.includes('.')) {
      const parts = trKey.split('.', 2)
      // TODO: maybe add typecheck
      const localization = this.getLocalizationElement(this.localization, parts[0]) as TSunCardI18N
      return this.getLocalizationElement(localization, parts[1])
    } else {
      // if the translation isn't completed in the target language, fall back to english
      // give ugly string for developers who misstype
      return (localization ? localization[trKey] : undefined)
        ?? Constants.FALLBACK_LOCALIZATION[trKey]
        ?? `Translation key '${trKey}' doesn't have a valid translation`
    }
  }
}

export default {
  nothing: html``,
  renderFieldElement (i18n: I18N, translationKey: string, value: Date | number | string): TemplateResult {
    
    let display: string
    if (value instanceof Date) {
      display = i18n.formatDateAsTime(value)
    } else {
      display = value.toString()
    }

    return html`
      <div class="sun-card-text-container">
        <span class="sun-card-field-name">${ i18n.tr(translationKey) }</span>
        <span class="sun-card-field-value">${ display }</span>
      </div>
    `
  },
  isValidLanguage (language: string): boolean {
    return Object.keys(Constants.LOCALIZATION_LANGUAGES).includes(language)
  },
  todayAtStartOfDay (): Date {
    const today = new Date()
    today.setHours(0)
    today.setMinutes(0)
    today.setSeconds(0)
    today.setMilliseconds(0)
  
    return today
  },
  todayAtEndOfDay (): Date {
    const today = new Date()
    today.setHours(23)
    today.setMinutes(59)
    today.setSeconds(59)
    today.setMilliseconds(999)
  
    return today
  },
  findSectionPosition (msSinceSectionStart: number, msSectionEnd: number, section: number): number {    
    return (Math.min(msSinceSectionStart, msSectionEnd) * section) / msSectionEnd
  },
  findSunProgress (sunPosition: number, startPosition: number, endPosition: number): number {
    return this.clamp(0, 100,
      (100 * (sunPosition - startPosition)) / (endPosition - startPosition)
    )
  },
  clamp (min: number, max: number, value: number): number {
    return Math.min(Math.max(value, min), max)
  }
}
