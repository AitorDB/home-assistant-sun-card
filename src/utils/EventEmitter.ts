type TEventCallback <T, U extends CustomEvent<T> = CustomEvent<T>, V extends (event: U) => void = (event: U) => void> = V
type TEventList = Record<string, unknown>
type TEventName <T> = Extract<keyof T, string>

export abstract class EventEmitter <T extends TEventList> {
  target: EventTarget

  constructor () {
    this.target = new EventTarget()
  }

  on <U extends TEventName<T> = TEventName<T>, W extends CustomEvent<T[U]> = CustomEvent<T[U]>, V extends TEventCallback<T[U], W> = TEventCallback<T[U], W>> (eventName: U, listener: V): void {
    return this.target.addEventListener(eventName, listener as unknown as EventListener) // FIXME: Remove "as unknown"
  }

  once <U extends TEventName<T> = TEventName<T>, W extends CustomEvent<T[U]> = CustomEvent<T[U]>, V extends TEventCallback<T[U], W> = TEventCallback<T[U], W>> (eventName: U, listener: V): void {
    return this.target.addEventListener(eventName, listener as unknown as EventListener, { once: true }) // FIXME: Remove "as unknown"
  }

  off <U extends TEventName<T> = TEventName<T>, W extends CustomEvent<T[U]> = CustomEvent<T[U]>, V extends TEventCallback<T[U], W> = TEventCallback<T[U], W>> (eventName: U, listener: V): void {
    return this.target.removeEventListener(eventName, listener as unknown as EventListener) // FIXME: Remove "as unknown"
  }

  emit <U extends TEventName<T> = TEventName<T>> (eventName: U, detail: T[U]): boolean
  emit <U extends TEventName<T> = TEventName<T>, W extends CustomEvent<T[U]> = CustomEvent<T[U]>> (eventName: U, detail: W): boolean
  emit <U extends TEventName<T> = TEventName<T>, W extends CustomEvent<T[U]> = CustomEvent<T[U]>> (eventName: U, data: T[U] | W): boolean {
    if (data instanceof Event) {
      return this.target.dispatchEvent(data)
    }

    const event = new CustomEvent(eventName, { detail: data, cancelable: true })
    return this.target.dispatchEvent(event)
  }
}
