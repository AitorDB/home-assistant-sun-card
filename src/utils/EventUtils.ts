// eslint-disable-next-line @typescript-eslint/no-explicit-any
type TEventList = Record<string, any>
type TEventName <T> = Extract<keyof T, string>

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type TEventListener <T = any> = (event: T) => void

export class EventUtils <T extends TEventList = TEventList> {
  eventMap: Map<TEventName<T>, TEventListener[]> = new Map()

  on <U extends TEventName<T> = TEventName<T>> (eventName: U, listener: TEventListener<T[U]>): void {
    const eventListeners = this.eventMap.get(eventName) || []
    eventListeners.push(listener)
    this.eventMap.set(eventName, eventListeners)
  }

  emit <U extends TEventName<T> = TEventName<T>> (eventName: U, data: T[U]): void {
    const eventListeners = this.eventMap.get(eventName) || []
    eventListeners.forEach((eventListener) => {
      eventListener(data)
    })
  }
}
