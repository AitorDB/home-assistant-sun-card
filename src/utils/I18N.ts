import { Constants } from '../constants'
import { TSunCardI18N, TSunCardI18NKeys } from '../types'

export class I18N {
  private localization: TSunCardI18NKeys
  private dateFormatter: Intl.DateTimeFormat

  constructor (language: string, use12HourClock: boolean | undefined) {
    this.localization = Constants.LOCALIZATION_LANGUAGES[language]

    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/DateTimeFormat
    const dateTimeFormatOptions: Intl.DateTimeFormatOptions = {
      timeStyle: 'short'
    }

    // if user hasn't defined this specifically in config
    // let the formatter figure it out based on language
    if (use12HourClock !== undefined) {
      dateTimeFormatOptions.hour12 = use12HourClock
    }

    this.dateFormatter = new Intl.DateTimeFormat(language, dateTimeFormatOptions)
  }

  public formatDateAsTime (date: Date): string {
    return this.dateFormatter.formatToParts(date).map(({ type, value }) => {
        switch (type) {
          // intentional fallthrough
          case 'hour':
          case 'minute':
          case 'dayPeriod':
          case 'literal':
            return value
          default:
            return ''
        }
      })
      .join('')
  }

  /**
   * TR -> TRanslation
   * @param translationKey The key to lookup a translation for
   * @returns The string specified in the translation files
   */
  public tr (translationKey: string): string {
    return this.getLocalizationElement(this.localization, translationKey).toString()
  }

  // Janky recursive logic to handle nested values in i18n json sources
  private getLocalizationElement (localization: TSunCardI18N, translationKey: string): string | TSunCardI18N {
    if (translationKey.includes('.')) {
      const parts = translationKey.split('.', 2)
      // TODO: maybe add typecheck
      const localization = this.getLocalizationElement(this.localization, parts[0]) as TSunCardI18N
      return this.getLocalizationElement(localization, parts[1])
    } else {
      // if the translation isn't completed in the target language, fall back to english
      // give ugly string for developers who misstype
      return (localization ? localization[translationKey] : undefined)
        ?? Constants.FALLBACK_LOCALIZATION[translationKey]
        ?? `Translation key '${translationKey}' doesn't have a valid translation`
    }
  }
}
