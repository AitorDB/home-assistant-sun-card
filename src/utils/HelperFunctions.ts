import { html, TemplateResult } from 'lit-html'
import { Constants } from '../constants'
import { I18N } from './I18N'

export class HelperFunctions {
  public static nothing (): TemplateResult {
    return html``
  }

  public static renderFieldElement (i18n: I18N, translationKey: string, value: Date | number | string | undefined): TemplateResult {
    let display: string
    if (value === undefined) {
      return HelperFunctions.nothing()
    } else if (value instanceof Date) {
      display = i18n.formatDateAsTime(value)
    } else {
      display = value.toString()
    }

    return html`
      <div class="sun-card-text-container">
        <span class="sun-card-field-name">${ i18n.tr(translationKey) }</span>
        <span class="sun-card-field-value">${ display }</span>
      </div>
    `
  }
  
  public static isValidLanguage (language: string): boolean {
    return Object.keys(Constants.LOCALIZATION_LANGUAGES).includes(language)
  }

  public static todayAtStartOfDay (): Date {
    const today = new Date()
    today.setHours(0)
    today.setMinutes(0)
    today.setSeconds(0)
    today.setMilliseconds(0)
  
    return today
  }
  
  public static todayAtEndOfDay (): Date {
    const today = new Date()
    today.setHours(23)
    today.setMinutes(59)
    today.setSeconds(59)
    today.setMilliseconds(999)
  
    return today
  }

  public static findSectionPosition (msSinceSectionStart: number, msSectionEnd: number, section: number): number {    
    return (Math.min(msSinceSectionStart, msSectionEnd) * section) / msSectionEnd
  }

  public static findSunProgress (sunPosition: number, startPosition: number, endPosition: number): number {
    return HelperFunctions.clamp(0, 100,
      (100 * (sunPosition - startPosition)) / (endPosition - startPosition)
    )
  }
  public static clamp (min: number, max: number, value: number): number {
    if (min === max) {
      return min
    }

    if (min > max) {
      throw new RangeError('Min value can not be bigger than the max value')
    }

    return Math.min(Math.max(value, min), max)
  }
}
