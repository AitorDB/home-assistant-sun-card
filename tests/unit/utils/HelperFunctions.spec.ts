import { HelperFunctions } from '../../../src/utils/HelperFunctions'
import { I18N } from '../../../src/utils/I18N'
import { CustomSnapshotSerializer, TemplateResultTestHelper } from '../../helpers/TestHelpers'

jest.mock('../../../src/utils/I18N', () => require('../../mocks/I18N'))

expect.addSnapshotSerializer(new CustomSnapshotSerializer())

describe('HelperFunctions', () => {
  describe('nothing', () => {
    it('returns an empty template result', async () => {
      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof HelperFunctions.nothing>
      element.templateResultFunction = HelperFunctions.nothing
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })
  })

  describe('renderFieldElement', () => {
    it('returns an empty template when the provided value is undefined', async () => {
      const i18 = new I18N('es', false)

      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof HelperFunctions.renderFieldElement>
      element.templateResultFunction = HelperFunctions.renderFieldElement
      element.templateResultFunctionData = [i18, 'sunrise', undefined]
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })

    it('returns a field element template when the provided value is a date', async () => {
      const i18 = new I18N('es', false)
      const date = new Date('2021-06-07T01:37:17.812Z')

      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof HelperFunctions.renderFieldElement>
      element.templateResultFunction = HelperFunctions.renderFieldElement
      element.templateResultFunctionData = [i18, 'sunrise', date]
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })

    it('returns a field element template when the provided value is a string', async () => {
      const i18 = new I18N('es', false)

      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof HelperFunctions.renderFieldElement>
      element.templateResultFunction = HelperFunctions.renderFieldElement
      element.templateResultFunctionData = [i18, 'sunrise', 'test']
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })

    it('returns a field element template when the provided value is a number', async () => {
      const i18 = new I18N('es', false)

      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof HelperFunctions.renderFieldElement>
      element.templateResultFunction = HelperFunctions.renderFieldElement
      element.templateResultFunctionData = [i18, 'sunrise', 9]
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })
  })

  describe('isValidLanguage', () => {
    it('returns true the provided language is supported', () => {
      const result = HelperFunctions.isValidLanguage('es')
      expect(result).toBe(true)
    })

    it('returns false the provided language is not supported', () => {
      const result = HelperFunctions.isValidLanguage('notSupported')
      expect(result).toBe(false)
    })
  })

  describe('todayAtStartOfDay', () => {
    it('returns today at the beginning of the day', () => {
      const result = HelperFunctions.todayAtStartOfDay(new Date(0))
      expect(result.getHours()).toBe(0)
      expect(result.getMinutes()).toBe(0)
      expect(result.getSeconds()).toBe(0)
      expect(result.getMilliseconds()).toBe(0)
    })
  })

  describe('todayAtEndOfDay', () => {
    it('returns today at the end of the day', () => {
      const result = HelperFunctions.todayAtEndOfDay(new Date(0))
      expect(result.getHours()).toBe(23)
      expect(result.getMinutes()).toBe(59)
      expect(result.getSeconds()).toBe(59)
      expect(result.getMilliseconds()).toBe(999)
    })
  })

  describe('findSectionPosition', () => {
    it('returns the correct section position when it has just started', () => {
      const timeSinceSectionStarted = 0
      const timeWhenSectionFinishes = 400
      const sectionSize = 150
      const result = HelperFunctions.findSectionPosition(timeSinceSectionStarted, timeWhenSectionFinishes, sectionSize)
      expect(result).toBe(0)
    })

    it('returns the correct section position when it has just finished', () => {
      const timeSinceSectionStarted = 400
      const timeWhenSectionFinishes = 400
      const sectionSize = 150
      const result = HelperFunctions.findSectionPosition(timeSinceSectionStarted, timeWhenSectionFinishes, sectionSize)
      expect(result).toBe(150)
    })

    it('returns the correct section position when it has finished', () => {
      const timeSinceSectionStarted = 600
      const timeWhenSectionFinishes = 400
      const sectionSize = 150
      const result = HelperFunctions.findSectionPosition(timeSinceSectionStarted, timeWhenSectionFinishes, sectionSize)
      expect(result).toBe(150)
    })
  })

  describe('findSunProgress', () => {
    it('returns the correct sun progress when the sun position is lower than the start position', () => {
      const sunPosition = 0
      const startPosition = 50
      const endPosition = 100
      const result = HelperFunctions.findSunProgress(sunPosition, startPosition, endPosition)
      expect(result).toBe(0)
    })

    it('returns the correct sun progress when the sun position is higher than the end position', () => {
      const sunPosition = 200
      const startPosition = 0
      const endPosition = 50
      const result = HelperFunctions.findSunProgress(sunPosition, startPosition, endPosition)
      expect(result).toBe(100)
    })

    it('returns the correct sun progress when the sun position is between the start and the end position', () => {
      const sunPosition = 20
      const startPosition = 0
      const endPosition = 100
      const result = HelperFunctions.findSunProgress(sunPosition, startPosition, endPosition)
      expect(result).toBe(20)
    })
  })

  describe('clamp', () => {
    it('returns the correct value when min and max have the same value', () => {
      const result = HelperFunctions.clamp(50, 50, 20)
      expect(result).toBe(50)
    })

    it('throws an error when min is greater than max', () => {
      let resultError
      try {
        HelperFunctions.clamp(50, 0, 20)
      } catch (error) {
        resultError = error
      }

      expect(resultError).toBeDefined()
      expect(resultError.name).toBe('RangeError')
      expect(resultError.message).toBe('Min value can not be bigger than the max value')
    })

    it('returns the correct value when the value is between min and max', () => {
      const result = HelperFunctions.clamp(0, 50, 20)
      expect(result).toBe(20)
    })

    it('returns the correct value when the value is lower than min', () => {
      const result = HelperFunctions.clamp(40, 50, 20)
      expect(result).toBe(40)
    })

    it('returns the correct value when the value is higher than max', () => {
      const result = HelperFunctions.clamp(40, 50, 60)
      expect(result).toBe(50)
    })
  })
})
