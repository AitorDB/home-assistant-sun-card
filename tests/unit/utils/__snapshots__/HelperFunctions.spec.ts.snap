// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelperFunctions nothing returns an empty template result 1`] = `"<!----><!---->"`;

exports[`HelperFunctions renderFieldElement returns a field element template when the provided value is a date 1`] = `
"<!---->
      <div class=\\"sun-card-text-container\\">
        <span class=\\"sun-card-field-name\\"><!---->sunrise<!----></span>
        <span class=\\"sun-card-field-value\\"><!---->2:37<!----></span>
      </div>
    <!---->"
`;

exports[`HelperFunctions renderFieldElement returns a field element template when the provided value is a number 1`] = `
"<!---->
      <div class=\\"sun-card-text-container\\">
        <span class=\\"sun-card-field-name\\"><!---->sunrise<!----></span>
        <span class=\\"sun-card-field-value\\"><!---->9<!----></span>
      </div>
    <!---->"
`;

exports[`HelperFunctions renderFieldElement returns a field element template when the provided value is a string 1`] = `
"<!---->
      <div class=\\"sun-card-text-container\\">
        <span class=\\"sun-card-field-name\\"><!---->sunrise<!----></span>
        <span class=\\"sun-card-field-value\\"><!---->test<!----></span>
      </div>
    <!---->"
`;

exports[`HelperFunctions renderFieldElement returns an empty template when the provided value is undefined 1`] = `"<!----><!---->"`;
